tracer/src/Datadog.Trace/AppSec/Security.cs

```diff
@@ -299,40 +302,58 @@ private bool UpdateRulesData()
 
         private void UpdateRuleStatus(IDictionary<string, bool> ruleStatus)
         {
-            if (ruleStatus != null && ruleStatus.Count > 0 && !_waf.ToggleRules(ruleStatus))
+            if (ruleStatus is { Count: > 0 } && !(_waf?.ToggleRules(ruleStatus) ?? false))
             {
                 Log.Debug($"_waf.ToggleRules returned false ({ruleStatus.Count} rule status entries)");
             }
         }
 
         private void UpdateStatus(bool fromRemoteConfig = false)
         {
-            lock (_settings)
+            if (_settings.Enabled)
             {
-                if (_settings.Enabled)
+                if (_libraryInitializationResult == null)
                 {
-                    _waf?.Dispose();
-
-                    _waf = Waf.Waf.Create(_settings.ObfuscationParameterKeyRegex, _settings.ObfuscationParameterValueRegex, _settings.Rules, _remoteRulesJson);
-                    if (_waf?.InitializedSuccessfully ?? false)
-                    {
-                        UpdateRulesData();
-                        EnableWaf(fromRemoteConfig);
-                    }
-                    else
+                    _libraryInitializationResult = WafLibraryInvoker.Initialize();
+                    if (!_libraryInitializationResult.Success)
                     {
                         _settings.Enabled = false;
+                        // logs happened during the process of initializing
+                        return;
                     }
+
+                    _wafLibraryInvoker = _libraryInitializationResult.WafLibraryInvoker;
                 }
 
-                if (!_settings.Enabled)
+                _wafInitializationResult = Waf.Waf.Create(_wafLibraryInvoker, _settings.ObfuscationParameterKeyRegex, _settings.ObfuscationParameterValueRegex, _settings.Rules, _remoteRulesJson);
+                if (_wafInitializationResult.Success && _wafLocker.EnterWriteLock())
```

Comment:
very good catch, because we werent disposing the new waf either ! fixed in [here like this](https://github.com/DataDog/dd-trace-dotnet/blob/359a06d7be571b81501689930888ed3ebfaa2f68/tracer/src/Datadog.Trace/AppSec/Security.cs#L331)
if we didnt acquire the lock, we just keep the older waf on and log a warning message for now